File: 152-maximum-product-subarray/152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int prod = 1;
        int maxProd = nums[0];
        int maxP,minP;
        maxP = nums[0];
        minP = nums[0];
        for( int i = 1; i < nums.size(); i++)
        {
            if(nums[i]<0)
                swap(maxP,minP);
            maxP = max(nums[i]*maxP,nums[i]);
            minP = min(nums[i]*minP,nums[i]);
            maxProd = max(maxP,maxProd);
        }
        return maxProd;
    }
};

File: 152-maximum-product-subarray/152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int maxProduct = Integer.MIN_VALUE;
        int currProduct = 1;
        for (int i =0;i<nums.length;i++) {
            currProduct *= nums[i];
            maxProduct = Math.max(currProduct, maxProduct);
            if (currProduct == 0) {
                currProduct = 1;
            }
        }
        currProduct = 1;
        for (int i = nums.length-1;i>=0;i--) {
            currProduct *= nums[i];
            maxProduct = Math.max(currProduct, maxProduct);
            if (currProduct == 0) {
                currProduct = 1;
            }
        }
        return maxProduct;
    }
}